rest/blip_sync.go

```diff
@@ -932,29 +934,36 @@ func (bh *blipHandler) handleRev(rq *blip.Message) error {
 			deltaSrcBody[db.BodyAttachments] = map[string]interface{}(deltaSrcRev.Attachments)
 		}
 
-		deltaSrcMap := map[string]interface{}(deltaSrcBody)
-		err = base.Patch(&deltaSrcMap, newDoc.Body())
+		var body db.Body
+
+		// Unmarshal delta including properties for patching
+		err = rq.ReadJSONBody(&body)
+		if err != nil {
+			return base.HTTPErrorf(http.StatusBadRequest, "error unmarshalling body: %v", err)
+		}
+
+		patchedBody := map[string]interface{}(deltaSrcBody)
+		err = base.Patch(&patchedBody, body)
 		if err != nil {
 			return base.HTTPErrorf(http.StatusInternalServerError, "Error patching deltaSrc with delta: %s", err)
 		}
 
-		newDoc.UpdateBody(deltaSrcMap)
-		bh.Logf(base.LevelTrace, base.KeySync, "docID: %s - body after patching: %v", base.UD(docID), base.UD(deltaSrcMap))
+		bh.Logf(base.LevelTrace, base.KeySync, "docID: %s - body after patching: %v", base.UD(docID), base.UD(patchedBody))
 		bh.db.DbStats.StatsDeltaSync().Add(base.StatKeyDeltaPushDocCount, 1)
-	}
 
-	// Handle and pull out expiry
-	if bytes.Contains(bodyBytes, []byte(db.BodyExpiry)) {
-		body := newDoc.Body()
-		expiry, err := body.ExtractExpiry()
+		err = newDoc.UpdateBody(patchedBody)
```

Comment:
We're doing a **_lot_** of work when a delta contains an attachment (probably not too uncommon a scenario)

1. incoming rev as `[]byte`
2. delta as `body`
3. do patch
4. updateBody to `[]byte`
5. attachment GetBody `body`
6. attachment ToIncomingDoc `[]byte`

We can probably avoid by adding some special handling for this scenario:

Store `patchedBody` as a body scoped outside of isDelta, and don't call UpdateBody from there.
In the `hasAttachments || hasExpiry` block, if the body created above is not nil, use that instead of calling `newDoc.GetBody()`.

this reduces the processing down to

1. incoming rev as `[]byte`
2. delta as `body`
3. do patch
4. attachmentToIncomingDoc to `[]byte`

And in an `else if newBody != nil` block off the end of `hasAttachments || hasExpiry`, you can do your `UpdateBody` for the delta without attachment case.